clients/drcachesim/tests/cache_replacement_policy_unit_test.cpp

```diff
@@ -0,0 +1,109 @@
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+// Unit tests for cache replacement policies
+#include <iostream>
+#include <assert.h>
+#include "cache_replacement_policy_unit_test.h"
+#include "simulator/cache_lru.h"
+
+class cache_lru_test_t : public cache_lru_t {
+public:
+    void
+    unit_test_replace_which_way()
+    {
+        initialize_cache(/*associativity=*/4, /*line_size=*/32, /*total_size=*/256);
+
+        // Access the cache line in the following fashion. This sequence follows the
+        // sequence shown in https://github.com/DynamoRIO/dynamorio/issues/4881.
+        const int ADDRESS_A = 0;
+        const int ADDRESS_B = 64;
+        const int ADDRESS_C = 128;
+        const int ADDRESS_D = 192;
+        const int ADDRESS_E = 72;
+
+        assert(get_block_index(ADDRESS_A) == get_block_index(ADDRESS_B));
+        assert(get_block_index(ADDRESS_B) == get_block_index(ADDRESS_C));
+        assert(get_block_index(ADDRESS_C) == get_block_index(ADDRESS_D));
+        assert(get_block_index(ADDRESS_D) == get_block_index(ADDRESS_E));
+
+        // Lower-case letter shows the least recently used way.
+        access_and_check_lru(ADDRESS_A, 1); // A X X X
+        access_and_check_lru(ADDRESS_B, 2); // A B X X
+        access_and_check_lru(ADDRESS_C, 3); // A B C X
+        access_and_check_lru(ADDRESS_D, 0); // a B C D
+        access_and_check_lru(ADDRESS_A, 1); // A b C D
+        access_and_check_lru(ADDRESS_A, 1); // A b C D
+        access_and_check_lru(ADDRESS_A, 1); // A b C D
+        access_and_check_lru(ADDRESS_E, 2); // A E c D
+
+        // XXX i#4842: Add more test sequences.
```

Comment:
It's not clear to me how new tests are going to fit in this existing code: they'll want a brand-new cache, or to invalidate the current one.  Their tests should probably be in separate functions.  Think about how to better structure this, probably renaming this function to reflect what it focuses on and moving the initialize_cache() to be called just once or something?  Or moving the test code to not be inside a cache instance so we can easily create a new cache for each test sequence: does this test code only need public functions?  What is your plan for more tests?

Reply:
Restructured it. Moved the test sequence outside of `cache_lru_test_t` class. That class is standalone and can be instantiated by caches of varying sizes implemented outside similar to `unit_test_cache_lru_four_way()`.